js/src/ipc/writer/binary.ts

```diff
@@ -0,0 +1,705 @@
+    return serializeMessage(dbMeta, rbData);
+}
+
+export function serializeMessage(message: Message, data?: Uint8Array) {
+    const b = new Builder();
+    _Message.finishMessageBuffer(b, writeMessage(b, message));
+    // Slice out the buffer that contains the message metadata
+    const metadataBytes = b.asUint8Array();
+    // Reserve 4 bytes for writing the message size at the front.
+    // Metadata length includes the metadata byteLength + the 4
+    // bytes for the length, and rounded up to the nearest 8 bytes.
+    const metadataLength = align(PADDING + metadataBytes.byteLength, 8);
+    // + the length of the optional data buffer at the end, padded
+    const dataByteLength = data ? data.byteLength : 0;
+    // ensure the entire message is aligned to an 8-byte boundary
+    const messageBytes = new Uint8Array(align(metadataLength + dataByteLength, 8));
+    // Write the metadata length into the first 4 bytes, but subtract the
+    // bytes we use to hold the length itself.
+    new DataView(messageBytes.buffer).setInt32(0, metadataLength - PADDING, platformIsLittleEndian);
+    // Copy the metadata bytes into the message buffer
+    messageBytes.set(metadataBytes, PADDING);
+    // Copy the optional data buffer after the metadata bytes
+    (data && dataByteLength > 0) && messageBytes.set(data, metadataLength);
+    // if (messageBytes.byteLength % 8 !== 0) { debugger; }
+    // Return the metadata length because we need to write it into each FileBlock also
+    return { metadataLength, buffer: messageBytes };
+}
+
+export function serializeFooter(footer: Footer) {
+    const b = new Builder();
+    _Footer.finishFooterBuffer(b, writeFooter(b, footer));
+    // Slice out the buffer that contains the footer metadata
+    const footerBytes = b.asUint8Array();
+    const metadataLength = footerBytes.byteLength;
+    return { metadataLength, buffer: footerBytes };
+}
+
+export class RecordBatchSerializer extends VectorVisitor {
+    public byteLength = 0;
+    public buffers: TypedArray[] = [];
+    public fieldNodes: FieldMetadata[] = [];
+    public buffersMeta: BufferMetadata[] = [];
+    public visitRecordBatch(recordBatch: RecordBatch) {
+        this.buffers = [];
+        this.byteLength = 0;
+        this.fieldNodes = [];
+        this.buffersMeta = [];
+        for (let vector: Vector, index = -1, numCols = recordBatch.numCols; ++index < numCols;) {
+            if (vector = recordBatch.getChildAt(index)!) {
+                this.visit(vector);
+            }
+        }
+        return this;
+    }
+    public visit<T extends DataType>(vector: Vector<T>) {
+        if (!DataType.isDictionary(vector.type)) {
+            const { data, length, nullCount } = vector;
+            if (length > 2147483647) {
+                throw new RangeError('Cannot write arrays larger than 2^31 - 1 in length');
+            }
+            this.fieldNodes.push(new FieldMetadata(length, nullCount));
+            this.addBuffer(nullCount <= 0
```

Comment:
Do negative null counts occur somehow in JS? We use `-1` in C++ to indicate that the null count has not been computed  (i.e. after an array slice has occurred)

Reply:
We technically [do the same thing as C++](https://github.com/apache/arrow/blob/fcc13f5088323989d925270e7850b817bde19d48/js/src/data.ts#L55) in the base ArrayData when someone slices a Vector, and [re-compute the `nullCount`](https://github.com/apache/arrow/blob/fcc13f5088323989d925270e7850b817bde19d48/js/src/data.ts#L85) the first time the getter is called after the slice, so here it'll be the real `nullCount` after any slices.

The kicker is the public `nullCount` is _almost definitely_ accessed immediately after [in the base Vector constructor](https://github.com/apache/arrow/blob/master/js/src/vector.ts#L53), so the optimization is ultimately moot. We've left it in here so we have a starting point to optimize in the future.